vllm/compilation/fusion_attn.py

```diff
@@ -0,0 +1,146 @@
+            attn_output_buf = attn_output_view.args[0]
+            assert is_func(attn_output_buf, EMPTY_OP)
+
+            # Check that the attention layer impl supports quant fusion
+            attn_layer: Attention = self.static_fwd_context[layer_name]
+            quant_dtype = quant_output.kwargs['dtype']
+            # hardcoded because we only match the op for static/per-tensor
+            # (static_scaled_fp8_quant) for now
+            static, per_token = True, False
+            if not attn_layer.impl.fused_output_quant_supported(
+                    quant_dtype, static=static, per_token=per_token):
+                continue
+
+            # The output shapes and dtypes are:
+            # - attn_view: bf16[s0, n_heads, head_size]
+            # - attn_buf: bf16[s0, n_heads * head_size]
+            # - quant: fp8[s0, n_heads * head_size]
+
+            # START REWRITE
+            # After this point, all operations must succeed for the graph to
+            # remain valid. That includes modifications to attention layers.
+
+            # 1. Set scale on unified_attn
+            node.update_kwarg('output_scale', quant_node.kwargs['scale'])
+
+            # 2. Rewrite the initial buf alloc dtype
+            attn_output_buf.update_kwarg('dtype', quant_output.kwargs['dtype'])
+            assert attn_output_buf.args == quant_output.args
+            assert attn_output_buf.kwargs == quant_output.kwargs
+            # meta value fixed by FakeTensorUpdater
+
+            # 3. Update the autofunc node meta, as the FakeTensorUpdater does
+            # not touch them. To get both the shape and type correct, we have
+            # to create the type manually, which is a bit ugly.
+            old_fake = node.meta['val'][1]
+            new_fake_tensor = torch.empty_like(old_fake, dtype=quant_dtype)
+            node.meta['val'] = (None, new_fake_tensor)
+
+            # 4. Autofunc nodes don't get updated, so add the next node to the
+            # update list. Also remove the meta value as the FakeTensorUpdater
+            # has a bug where the type is not considered in FakeTensor equality.
+            # No existing meta val forces a new one.
+            nodes_to_process.append(attn_getitem)
+            del attn_getitem.meta['val']
+
+            # 5. Reshape output buffer back to 2d for scaled mm input
+            with graph.inserting_after(attn_getitem):
+                reshape_args = (attn_getitem, quant_output_shape)
+                attn_new_output = graph.call_function(RESHAPE_OP, reshape_args)
+
+            # 6. Rebind users of quant to use output of attention directly:
+            find_getitem(quant_node, 1).replace_all_uses_with(attn_new_output)
+
+            # Done, add to the count
+            count += 1
+
+        # remove the quant nodes
+        graph.eliminate_dead_code()
+
+        # Manually force processing of nodes
+        for node in nodes_to_process:
```

Comment:
Why do we need to do this ?

Reply:
Later passes rely on accurate meta information, and because we update nodes manually here, it's not set. Also, this gets run anyway after the post pass and fails if not correct. Due to bugs described above (not propagating through auto-func and sometimes when only dtype changes), it will fail in the `FakeTensorUpdater` called by upstream torch.compile.